pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -809,22 +808,25 @@ func (a *AmbientIndex) handleService(obj any, isDelete bool, c *Controller) map[
 	// Fetch updates again, in case it changed from adding new workloads
 	for _, vip := range vips {
 		for _, wl := range a.byService[vip] {
-			updates[model.ConfigKey{Kind: kind.Address, Name: wl.ResourceName()}] = struct{}{}
+			updates[model.ConfigKey{Kind: kind.Address, Name: wl.ResourceName(), Namespace: wl.ClusterId}] = struct{}{}
 		}
 	}
 	return updates
 }
 
 // PodInformation returns all WorkloadInfo's in the cluster.
 // This may be scoped to specific subsets by specifying a non-empty addresses field
-func (c *Controller) PodInformation(addresses sets.Set[types.NamespacedName]) ([]*model.WorkloadInfo, []string) {
+func (c *Controller) WorkloadInfos(addresses sets.Set[types.NamespacedName]) ([]*model.WorkloadInfo, []string) {
 	if len(addresses) == 0 {
 		// Full update
 		return c.ambientIndex.All(), nil
 	}
 	var wls []*model.WorkloadInfo
 	var removed []string
 	for p := range addresses {
+		if p.Namespace != "" && p.Namespace != c.Cluster().String() {
```

Comment:
Using clusterID we can check directly

But using networkID, we need to call the function to calculate per address

Reply:
Calculate what?

Reply:
```
// return the mesh network for the endpoint IP. Empty string if not found.
func (b *EndpointBuilder) endpointNetwork(endpointIP string) network.ID {
	// If we're building the endpoint based on proxy meta, prefer the injected ISTIO_META_NETWORK value.
	if b.metaNetwork != "" {
		return b.metaNetwork
	}

	return b.controller.Network(endpointIP, b.labels)
}
```

Reply:
That calculation is pretty cheap, its just a few conditional lookups. If looking up a network is slow we should optimize that, not break the model.

Addresses are scoped by network, not cluster.

I recommend we bundle making Address network-aware with making it WorkloadEntry (https://github.com/istio/istio/pull/43799/files) so we don't have overlapping/wasted work

Reply:
done